shark_turbine/kernel/lang/functional_types.py

```diff
@@ -0,0 +1,137 @@
+from typing import Optional, Type, TypeVar, cast, ClassVar
+from enum import Enum
+
+from ..lang.kernel_buffer import KernelBufferUsage
+from ..ops.wave_ops import register
+from .._support.shaped_type import ShapedDataType
+from .._support.dtype import DataType
+from ...kernel._support.indexing import IndexExpr
+import sympy
+import torch
+import torch.fx
+
+__all__ = [
+    "Memory",
+    "Register",
+    "AddressSpace",
+]
+
+MemoryTypeT = TypeVar("MemoryTypeT")
+
+
+class AddressSpace(Enum):
+    REGISTER = 0
+    SHARED_MEMORY = 1
+    GLOBAL_MEMORY = 2
+
+
+class _MemoryStorage(ShapedDataType):
+    def new_subtype(
+        cls: Type[MemoryTypeT],
+        *,
+        symbolic_shape: tuple[IndexExpr, ...],
+        address_space: AddressSpace,
+        dtype: DataType,
+        usage: Optional[KernelBufferUsage] = None,
+    ) -> Type[MemoryTypeT]:
+        init_symbolic_shape = symbolic_shape
+        init_dtype = dtype
+        init_address_space = (
+            address_space if address_space else AddressSpace.REGISTER.value
+        )
+        init_usage = usage
+
+        class MemoryType(cls):
+            symbolic_shape = init_symbolic_shape
+            rank = len(symbolic_shape)
+            address_space = init_address_space
+            dtype = init_dtype
+            usage = init_usage
+
+        return cast(Type[MemoryTypeT], MemoryType)
+
+
+class Memory(metaclass=_MemoryStorage):
```

Comment:
Is this type different from KernelBuffer? If not, can we reuse them?

Reply:
I believe the biggest difference is the addition of the AddressSpace in the type to be able to model different levels in the memory hierarchy (think AIE for example which has multiple levels) and also didn't want to break things in the other pipeline but will let Martin comment as well.

Reply:
Indeed, the plan is to infer where the memory is to be placed and also whether it is used for input / output / both.
Examples of the supported usage is [here](https://github.com/iree-org/iree-turbine/blob/eaa04e20389df28304eeb3efe12acc8c68738ca8/tests/kernel/wave/types_test.py#L15)

Reply:
It's fine to land this, but effectively, KernelBuffer and this MemoryStorage are the same thing. KernelBufferUsage.Local would just get another attribute specifying what kind it is.

Reply:
You are right, I reworked the types to build on top of `KernelBufferMeta` instead.
For that I changed it as follows:
- made it public
- added address_space to it + moved enum to this file
- added the option to manually specify a type name besides inferring from KernelBufferUsage

We should only deviate from the established framework if we have good reasons, so thanks for highlighting this.

Let me know if this is alright for you.